### PR TITLE
[spark] Add `load_spark_dataset` API

### PR DESCRIPTION
**Pull request by WeichenXu123**
Signed-off-by: Weichen Xu <weichen.xu@databricks.com>

## Why are these changes needed?

Add `load_spark_dataset` API.
This API can be used to load data from spark dataset easily.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/util/spark/data.py
Comment (jjyao) [inline python/ray/util/spark/data.py:12]: Shouldn't this be added to Ray Dataset API directly? cc @clarkzinzow @c21 for discussion.

**Review by jjyao — CHANGES_REQUESTED** [review]

**Review by architkulkarni — APPROVED** [review]
Codeowner stamp for Cluster docs

File: python/ray/util/spark/data.py
Comment (WeichenXu123) [reply python/ray/util/spark/data.py:12]: This sounds like a good idea

File: python/ray/tests/spark/test_data.py
Comment (harupy) [inline python/ray/tests/spark/test_data.py:20]: You can use the `tmp_path` fixture.

File: python/ray/util/spark/data.py
Comment (harupy) [inline python/ray/util/spark/data.py:26]:
```suggestion
    assert saved_format in [
        "parquet",
        "delta",
    ], f"Unsupported saved format {saved_format}"
    spark = get_spark_session()
```

File: python/ray/tests/spark/test_data.py
Comment (harupy) [inline python/ray/tests/spark/test_data.py:14]:
```suggestion
    with SparkSession.builder.getOrCreate() as spark: 
        yield spark
```

Does SparkSession work as a context manager?

File: python/ray/util/spark/data.py
Comment (WeichenXu123) [reply python/ray/util/spark/data.py:26]: I am considering to automatically infer the saved data type.

File: python/ray/util/spark/data.py
Comment (WeichenXu123) [reply python/ray/util/spark/data.py:12]: CC @clarkzinzow @c21 Thoughts ?

Comment (ericl) [reply python/ray/util/spark/data.py:12]: Hmm, it seems like the main thing this is providing is delta compatibility. If we want to add it natively, it could be something like ray.data.read_delta().

For supporting dbfs:, I think that could be an thin arrow filesystem impl, so the final call could be ray.data.read_delta(..., filesystem=DatabricksFileSystem()). The filesystem could just translate paths and delegate to the local arrow filesystem.

Comment (WeichenXu123) [reply python/ray/util/spark/data.py:12]: >  it seems like the main thing this is providing is delta compatibility. If we want to add it natively, it could be something like ray.data.read_delta().

@ericl 

Delta format saved data are also parquet format files, (but with some additional logs).
So read parquets can directly support read delta data. Do we need another API to do that ?

For apache/spark, by default the DataFrame is saved in parquet format, 
but in databricks runtime, by default the DataFrame is saved in delta format, 

so if the Ray loader API can automatically support parquet/delta format, it will be more friendly to user, user don't need to check whether the saved data is parquet or delta format.

Comment (WeichenXu123) [reply python/ray/util/spark/data.py:12]: Btw, we also want to support load data from "spark table" via the same API,
i.e. the `load_spark_dataset(source)`, source can be either a spark dataset path (parquet or delta format), or a spark table name. 

In this case, shall we still move this method to `ray.data` ? Its interface is not like other APIs in `ray.data`.

**Comment by WeichenXu123** [comment]
I will discuss with Anyscale folks to determine the API first, then reopen the ticket.

**Pull request closed without merging**
